reference/src/clientimpl/src/sa_key_provision.c

```diff
@@ -0,0 +1,364 @@
+               NULL == ((NetflixProvisioning*)in)->esnContainer    ||
+               0 == ((NetflixProvisioning*)in)->esnContainerLength) {
+               ERROR("null input or input size is  0");
+               return SA_STATUS_NULL_PARAMETER;
+            }
+
+            nflxProvision = calloc(sizeof(NetflixProvisioning), 1);
+            if(NULL == nflxProvision) {
+               ERROR("OOM");
+               return SA_STATUS_INTERNAL_ERROR;
+            }
+
+            /*deep copy*/
+            nflxProvision->hmacKeyLength =
+               ((NetflixProvisioning*)in)->hmacKeyLength;
+            nflxProvision->wrappingKeyLength =
+               ((NetflixProvisioning*)in)->wrappingKeyLength;
+            nflxProvision->esnContainerLength =
+               ((NetflixProvisioning*)in)->esnContainerLength;
+
+            nflxProvision->hmacKey =
+               (void*)calloc(nflxProvision->hmacKeyLength, 1);
+            if(NULL == nflxProvision->hmacKey) {
+               ERROR("OOM");
+               return SA_STATUS_INTERNAL_ERROR;
+            }
+            memcpy(nflxProvision->hmacKey, ((NetflixProvisioning*)in)->hmacKey,
+                   nflxProvision->hmacKeyLength);
+
+            nflxProvision->wrappingKey =
+               (void*)calloc(nflxProvision->wrappingKeyLength, 1);
+            if(NULL == nflxProvision->wrappingKey) {
+               ERROR("OOM");
+               return SA_STATUS_INTERNAL_ERROR;
+            }
+            memcpy(nflxProvision->wrappingKey,
+                   ((NetflixProvisioning*)in)->wrappingKey,
+                   nflxProvision->wrappingKeyLength);
+
+            /*here just provide an examlpe to pass ESN data,
+             how to use it is all up toSOC vendors.
+            */
+            nflxProvision->esnContainer =
+               (void*)calloc(nflxProvision->esnContainerLength, 1);
+            if(NULL == nflxProvision->esnContainer) {
+               ERROR("OOM");
+               return SA_STATUS_INTERNAL_ERROR;
+            }
+            memcpy(nflxProvision->esnContainer,
+                   ((NetflixProvisioning*)in)->esnContainer,
+                   nflxProvision->esnContainerLength);
+
+            INFO("nflxProvision:%x",nflxProvision);
+            INFO("hmacLen : %d", nflxProvision->hmacKeyLength);
+            INFO("wrappingLen : %d", nflxProvision->wrappingKeyLength);
+            INFO("ESNLen : %d", nflxProvision->esnContainerLength);
+
+            sa_key *sa_hmac_key = create_uninitialized_sa_key();
+            assert(NULL != sa_hmac_key);
+            sa_status status_hmac = sa_key_import(sa_hmac_key, SA_KEY_FORMAT_EXPORTED,
+                                  nflxProvision->hmacKey,nflxProvision->hmacKeyLength,
+                                  NULL);
+
+            sa_key *sa_wrapping_key = create_uninitialized_sa_key();
+            assert(NULL != sa_wrapping_key);
+
+            sa_status status_wrapping = sa_key_import(sa_wrapping_key, SA_KEY_FORMAT_EXPORTED,
```

Comment:
Shouldn't the parameters field provide the SA_KEY_FORMAT_SOC type to denote the key is protected with the SOC key?